File: apps/mobile/src/ui/__tests__/modelLabel.test.ts
/**
 * CDX-087: compact model tags for the header badge, ported from the old app's
 * `modelLabel` (codedeck/src/constants/models.ts:49-55).
 */
import { describe, expect, it } from 'vitest';
import { modelLabel } from '../modelLabel';

describe('modelLabel', () => {
  it('tags the known models the way the old app did', () => {
    expect(modelLabel('claude-opus-5')).toBe('O5');
    expect(modelLabel('claude-opus-4-8')).toBe('O4.8');
    expect(modelLabel('claude-sonnet-4-6')).toBe('S4.6');
    expect(modelLabel('claude-haiku-4-5-20251001')).toBe('H4.5');
    expect(modelLabel('claude-fable-5')).toBe('F5');
  });

  it('strips the 1M-context marker before lookup', () => {
    // The old app had no row for these, so they fell through to the derived
    // branch and read `opus-5[1m]` in a badge with room for four characters.
    // The 1M window is already visible in the context figure beside the tag.
    expect(modelLabel('claude-opus-5[1m]')).toBe('O5');
    expect(modelLabel('claude-opus-4-8[1m]')).toBe('O4.8');
    expect(modelLabel('claude-opus-5-1m')).toBe('O5');
  });

  it('derives something usable for an unknown model', () => {
    // A new release, or a custom provider profile's id (CDX-062).
    expect(modelLabel('claude-opus-9')).toBe('opus-9');
    expect(modelLabel('claude-sonnet-7-20270101')).toBe('sonnet-7');
    expect(modelLabel('kimi-k3-turbo')).toBe('kimi-k3-turbo');
  });

  it('reports ? rather than guessing when no model is recorded', () => {
    expect(modelLabel(undefined)).toBe('?');
    expect(modelLabel('')).toBe('?');
  });
});

File: apps/android/app/src/test/java/com/codedeck/plus/ui/session/ModelLabelTest.kt
package com.codedeck.plus.ui.session

import org.junit.Assert.assertEquals
import org.junit.Test

/** Port of `apps/mobile/src/ui/__tests__/modelLabel.test.ts` — the same
 *  cases, asserting the Kotlin port of `modelLabel` (SessionScreen.kt) keeps
 *  the reference's exact display convention for the header's model tag. */
class ModelLabelTest {
    @Test
    fun tagsTheKnownModelsLikeTheReference() {
        assertEquals("O5", modelLabel("claude-opus-5"))
        assertEquals("O4.8", modelLabel("claude-opus-4-8"))
        assertEquals("S4.6", modelLabel("claude-sonnet-4-6"))
        assertEquals("H4.5", modelLabel("claude-haiku-4-5-20251001"))
        assertEquals("F5", modelLabel("claude-fable-5"))
    }

    @Test
    fun stripsTheOneMContextMarkerBeforeLookup() {
        // These fall through the tag table without the strip and would read
        // `opus-5[1m]` in a badge with room for a handful of characters; the
        // 1M window is already visible in the context figure beside the tag.
        assertEquals("O5", modelLabel("claude-opus-5[1m]"))
        assertEquals("O4.8", modelLabel("claude-opus-4-8[1m]"))
        assertEquals("O5", modelLabel("claude-opus-5-1m"))
    }

    @Test
    fun derivesSomethingUsableForAnUnknownModel() {
        // A new release, or a custom provider profile's id.
        assertEquals("opus-9", modelLabel("claude-opus-9"))
        assertEquals("sonnet-7", modelLabel("claude-sonnet-7-20270101"))
        // Non-claude ids pass through unchanged (the header's marquee, not
        // this function, is what keeps a long one legible).
        assertEquals("kimi-k3-turbo", modelLabel("kimi-k3-turbo"))
        assertEquals(
            "Z.ai (Global) - Coding Plan/glm-5.3-flash",
            modelLabel("Z.ai (Global) - Coding Plan/glm-5.3-flash"),
        )
    }

    @Test
    fun reportsQuestionMarkRatherThanGuessingWhenNoModelIsRecorded() {
        assertEquals("?", modelLabel(null))
        assertEquals("?", modelLabel(""))
    }
}
